### ISSUE TITLE
npm install should be scoped to workspaces

### ISSUE THREAD

**Issue by nathanhammond** [renovatebot/renovate#20469]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

This PR created a synthetic dependency that does not exist:

https://github.com/vercel/turbo/pull/3422/

I've drawn attention to it here in the generated lockfile:
https://github.com/vercel/turbo/pull/3422#pullrequestreview-1302789067

You can see that it does not exist at the merge-base:
https://github.com/vercel/turbo/blob/49f4d52a56d06e9f88a14aaf72e32f118a8a6094/examples/with-npm/package.json

### Have you created a minimal reproduction repository?

No reproduction, but I have linked to a public repo where it occurs

**Comment by rarkins** [comment]
Please locate the logs for the run when the commit was made. This will show the `npm` commands which were run to generate the commit

**Comment by nathanhammond** [comment]
@rarkins there is no way that I can search for the SHA `53fab937ef7236781f32fba9f982f2ceafab0edf` in the Renovate dashboard. The initial commit was created nearly a month ago and the UI provided for digging out logs is extremely primitive. There appears to be no way for me to dig the logs out past the first page of results, spanning just three days.

As a feature for you to improve your debugging experience the commit or the PR should capture the job ID in the commit message.

**Comment by rarkins** [comment]
In that case you can check the retry checkbox and the commit will be reapplied

**Comment by glsignal** [comment]
Jumping in as we've encountered this as well, for a number of our dependencies. Logs included below for one such run @rarkins

The screenshotted dependency is being listed erroneously in the dependencies of the workspace root, despite not being listed in the package

```
frontend/apps/financial-controlling/package.json (lists autonumeric as a dependency)

frontend/package.json (no autonumeric dependency)
frontend/package-lock.json (autonumeric incorrectly being added to the deps list)
```

(note the screenshot is replacing another dep in that list, this was added due to the same synthesis issue)

![image](https://user-images.githubusercontent.com/1109291/227205417-d75d32d0-22d9-48fb-a389-7c7f18e06c4c.png)

<details>
<summary>Logs</summary>

<pre>
DEBUG: syncBranchState()(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: getBranchPr(renovate/autonumeric-4.x-lockfile)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: findPr(renovate/autonumeric-4.x-lockfile, undefined, open)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: findPr(renovate/autonumeric-4.x-lockfile, undefined, closed)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: branchExists=false(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: dependencyDashboardCheck=undefined(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: recreateClosed is false(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: findPr(renovate/autonumeric-4.x-lockfile, Update dependency autonumeric to v4.7.0, !open)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: prAlreadyExisted=false(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Checking schedule(after 10pm every weekday,before 5am every weekday,every weekend, null)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Checking 3 schedule(s)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Checking schedule "after 10pm every weekday"(branch="renovate/autonumeric-4.x-lockfile")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_a": [
          79200
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Checking schedule "before 5am every weekday"(branch="renovate/autonumeric-4.x-lockfile")
{
  "parsedSchedule": {
    "schedules": [
      {
        "t_b": [
          18000
        ],
        "d": [
          2,
          3,
          4,
          5,
          6
        ]
      }
    ],
    "exceptions": [],
    "error": -1
  }
}
DEBUG: Matches schedule before 5am every weekday(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Branch needs creating(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Using reuseExistingBranch: false(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Setting current branch to master(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: latest commit(branch="renovate/autonumeric-4.x-lockfile")
{
  "branchName": "master",
  "latestCommitDate": "2023-03-21T18:04:35+01:00"
}
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: npm.updateLockedDependency: autonumeric@4.6.0 -> 4.7.0 [frontend/package-lock.json](branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Found matching dependencies with length 1(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: autonumeric can be updated to 4.7.0 in-range with matching constraint "^4.6.0" in frontend/apps/financial-controlling/package.json(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: autonumeric can be updated to 4.7.0 in-range with matching constraint "^4.6.0" in financial-controlling@file:apps/financial-controlling(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Remediations which change package-lock.json only are not supported unless lockfileVersion=1(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: No package files need updating(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Getting updated lock files(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Writing package.json files(branch="renovate/autonumeric-4.x-lockfile")
{
  "packageFiles": [
    "backend/alasco/components/public_api/docs/package.json",
    "frontend/apps/financial-controlling/package.json",
    "frontend/package.json"
  ]
}
DEBUG: Writing backend/alasco/components/public_api/docs/package-lock.json(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Writing frontend/package-lock.json(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Massaging npm lock file before writing to disk(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Writing frontend/package-lock.json(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Massaging npm lock file before writing to disk(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Writing any updated package files(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: npmrc file found in repository(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Writing updated .npmrc file to frontend/.npmrc(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Generating package-lock.json for frontend(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Spawning npm install to create frontend/package-lock.json(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Updating lock file only(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Using node constraint ">=16.13.0" from package.json(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Performing lockfileUpdate (npm)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Setting CONTAINERBASE_CACHE_DIR to /tmp/containerbase(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Using docker to execute(branch="renovate/autonumeric-4.x-lockfile")
{
  "image": "sidecar"
}
DEBUG: Resolved stable matching version(branch="renovate/autonumeric-4.x-lockfile")
{
  "toolName": "node",
  "constraint": ">=16.13.0",
  "resolvedVersion": "v18.15.0"
}
DEBUG: Resolved stable matching version(branch="renovate/autonumeric-4.x-lockfile")
{
  "toolName": "npm",
  "constraint": ">=8.0.0 <9",
  "resolvedVersion": "8.19.4"
}
DEBUG: containerbaseDir is separate from cacheDir(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Resolved tag constraint(branch="renovate/autonumeric-4.x-lockfile")
{
  "image": "docker.io/containerbase/sidecar"
}
DEBUG: Docker image is already prefetched: docker.io/containerbase/sidecar@sha256:2cf35b6530dcf59fd00ec9da371924e4bafdf0c531d604820d1a16869ad4fbc0(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Executing command(branch="renovate/autonumeric-4.x-lockfile")
{
  "command": "docker run --rm --name=renovate_sidecar --label=renovate_child -v \"/mnt/renovate/gh/alasco-tech/alasco-app\":\"/mnt/renovate/gh/alasco-tech/alasco-app\" -v \"/tmp/renovate-cache\":\"/tmp/renovate-cache\" -v \"/tmp/containerbase\":\"/tmp/containerbase\" -e NPM_CONFIG_CACHE -e npm_config_store -e BUILDPACK_CACHE_DIR -e CONTAINERBASE_CACHE_DIR -w \"/mnt/renovate/gh/alasco-tech/alasco-app/frontend\" docker.io/containerbase/sidecar bash -l -c \"install-tool node v18.15.0 && install-tool npm 8.19.4 && hash -d npm 2>/dev/null || true && npm install --package-lock-only --no-audit --ignore-scripts autonumeric@4.7.0\""
}
DEBUG: exec completed(branch="renovate/autonumeric-4.x-lockfile")
{
  "durationMs": 10186,
  "stdout": "installing v2 tool node v18.15.0\nlinking tool node v18.15.0\nnode: v18.15.0 /usr/local/bin/node\nnpm: 9.5.0  /usr/local/bin/npm\nInstalled v2 /usr/local/buildpack/tools/v2/node.sh in 3 seconds\ninstalling v2 tool npm v8.19.4\nlinking tool npm v8.19.4\n8.19.4\nInstalled v2 /usr/local/buildpack/tools/v2/npm.sh in 3 seconds\n\nup to date in 3s\n\n275 packages are looking for funding\n  run `npm fund` for details\n",
  "stderr": ""
}
DEBUG: frontend/package-lock.json needs updating(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Updated 1 lock files(branch="renovate/autonumeric-4.x-lockfile")
{
  "updatedArtifacts": [
    "frontend/package-lock.json"
  ]
}
DEBUG: 1 file(s) to commit(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Preparing files for committing to branch renovate/autonumeric-4.x-lockfile(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: git commit(branch="renovate/autonumeric-4.x-lockfile")
{
  "deletedFiles": [],
  "ignoredFiles": [],
  "result": {
    "author": null,
    "branch": "renovate/autonumeric-4.x-lockfile",
    "commit": "4d0efcabc0f7480f460f31c039a1cdd01f039d3e",
    "root": false,
    "summary": {
      "changes": 1,
      "insertions": 445,
      "deletions": 0
    }
  }
}
DEBUG: resetToCommit(62652452968354f636e19c5198868c4e66151287)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Fetching branch renovate/autonumeric-4.x-lockfile(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Setting current branch to master(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: latest commit(branch="renovate/autonumeric-4.x-lockfile")
{
  "branchName": "master",
  "latestCommitDate": "2023-03-21T18:04:35+01:00"
}
INFO: Branch created(branch="renovate/autonumeric-4.x-lockfile")
{
  "commitSha": "6f973496522d7a1598b1c3254dd5aae65019a374"
}
DEBUG: Ensuring PR(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: There are 0 errors and 0 warnings(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: getBranchPr(renovate/autonumeric-4.x-lockfile)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: findPr(renovate/autonumeric-4.x-lockfile, undefined, open)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: findPr(renovate/autonumeric-4.x-lockfile, undefined, closed)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: getPrCache()(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Fetching changelog: https://github.com/autoNumeric/autoNumeric (4.6.0 -> 4.7.0)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Creating PR(branch="renovate/autonumeric-4.x-lockfile")
{
  "prTitle": "Update dependency autonumeric to v4.7.0"
}
DEBUG: Creating PR(branch="renovate/autonumeric-4.x-lockfile")
{
  "title": "Update dependency autonumeric to v4.7.0",
  "head": "alasco-tech:renovate/autonumeric-4.x-lockfile",
  "base": "master",
  "draft": false
}
DEBUG: PR created(branch="renovate/autonumeric-4.x-lockfile")
{
  "pr": 9835,
  "draft": false
}
DEBUG: Adding labels 'dependencies, javascript' to #9835(branch="renovate/autonumeric-4.x-lockfile")
INFO: PR created(branch="renovate/autonumeric-4.x-lockfile")
{
  "pr": 9835,
  "prTitle": "Update dependency autonumeric to v4.7.0"
}
DEBUG: addParticipants(pr=9835)(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: setPrCache()(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: Created Pull Request #9835(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: PR is not configured for automerge(branch="renovate/autonumeric-4.x-lockfile")
DEBUG: setBranchCommit()(branch="renovate/autonumeric-4.x-lockfile")
</pre>

</details>

**Comment by rarkins** [comment]
Renovate doesn't run `npm install`, it runs a command such as `npm install --package-lock-only --no-audit --ignore-scripts autonumeric@4.7.0` (you can see it from the logs above). So perhaps the error is related to `npm` and `--package-lock-only`, for example. This would be the first step to confirm.

**Comment by glsignal** [comment]
That looks like the probable cause, running that indeed produced the same problem with the lockfile.

Does renovate currently have any handling related to workspaces with npm? If we issue the same command and specify the workspace it produces the expected result (i.e. `npm install --workspace=apps/financial-controlling --package-lock-only --no-audit --ignore-scripts autonumeric@4.7.0`)

**Comment by glsignal** [comment]
[Here's the repro](https://github.com/glsignal/renovate-npm-workspace-repro) which shows the issue

**Comment by rarkins** [comment]
Great! Thanks.

Do you think we should _always_ specify the workspace when running npm install, or might that break some other use cases? Either we can:
- always do it, or
- cleverly autodetect when to do it, or
- make it opt-in by config

The latter is less ideal obviously..

**Comment by glsignal** [comment]
To be completely honest, I'm relatively new to working in projects with workspaces so I don't have a good handle on what potential edge cases might exist if it's always specified. So, please take this with a heavy pinch of salt!

Depending on what 'clever' refers to, the "always do it" approach sounds to me at least easier to understand (and debug) in case something _does_ go wrong, and at that point there would be more information to figure out which use cases need to be accounted for.

That said, it would least need to be able to distinguish between packages in the root workspace from nested ones, since (as far as I know) there isn't a way to specify explicitly that the workspace is the root one, rather it's the lack of `--workspace=blah` that does it

**Comment by rarkins** [comment]
Thanks, no problem.

npm workspaces docs: https://docs.npmjs.com/cli/v8/using-npm/workspaces

Seems like we need:
- To know if we're using workspaces
- To configure all workspaces which the install is for

I'm not sure what to do if the install is for a workspace *and* the root. Maybe we need two commands (one with workspace, one without)

**Comment by rarkins** [comment]
Reproduction forked to https://github.com/renovate-reproductions/20469

**Comment by rarkins** [comment]
This makes sense, please go ahead!

**Comment by RahulGautamSingh** [comment]
While working on implementation I stubmled upon how dependabot handles this:
It seems they have a fix but I couldn't reproduce this locally..when I try there method which I think is:
1. install normally once (like we currently do)
2. install without any arguments (ie. a normal installl without adding dep@version)

This didn't work in my case. Can you guys see if it works for you and maybe I missed something?

https://github.com/dependabot/dependabot-core/blob/50eadf172b8f13125c9ec6f963e5b595a8c642bc/npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb#L179-L183

**Comment by RahulGautamSingh** [comment]
>npm install --packagelock--only --no-audit --ignore-scripts --workspace=a A@x.x.x
npm install --packagelock--only --no-audit --ignore-scripts --workspace=b B@x.x.x
npm install --packagelock--only --no-audit --ignore-scripts --workspace=a --workspace=b D@x.x.x
npm install --packagelock--only --no-audit --ignore-scripts C@x.x.x E@x.x.x

npm install --packagelock--only --no-audit --ignore-scripts --workspace=a A@x.x.x D@x.x.x
npm install --packagelock--only --no-audit --ignore-scripts --workspace=b B@x.x.x D@x.x.x
npm install --packagelock--only --no-audit --ignore-scripts C@x.x.x E@x.x.x

Should do it this way. Efficient and less complex to code.

**Comment by RahulGautamSingh** [comment]
@nathanhammond I have tried fixing the https://github.com/renovatebot/renovate/pull/21274 and reproduced the issue you encountered https://github.com/RahulGautamSingh-testing/turbo. Could you confirm the issue is fixed?
_PS. I don't have pnpm installed locally thus the artifact issue, I think it can be ignored for this case._

**Comment by glsignal** [comment]
Thank you all for the fix ❤️

**Comment by batbattur** [comment]
Can we do the same for `pnpm`? Or do the changes from https://github.com/renovatebot/renovate/pull/21274, apply to pnpm as well?

**Comment by RahulGautamSingh** [comment]
> Can we do the same for `pnpm`? Or do the changes from #21274, apply to pnpm as well?

You can create a minimal reproduction and testif it works or not. If it doesn't then you can create a new issue and we can discuss the solution before starting work on it.
